### PR TITLE
Fixes selector logic of service and endpoint

### PR DESCRIPTION
**Pull request by miaoyq**
**Please provide a description of this PR:**

In the following code snippet, se.WorkloadSelector != nil, but se.Selector.WorkloadSelector is nil or an empty map, the se will select all workload instance.
https://github.com/istio/istio/blob/4157e1887071037b369fd82cb4559dc7628bb57b/pilot/pkg/serviceregistry/serviceentry/controller.go#L513-L521

However, in normal logic, any WorkloadInstance and Endpoint should not be selected when se's selector is empty.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Seems reasonable. Is this consistent across all types?

for example
```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: allow-nothing
 namespace: istio-system
spec:
 selector:
   matchLabels: {}
```

**Comment by howardjohn** [comment]
But we don't say how the others like ServiceEntry should behave so I am a
bit worried about the ambiguity there...
https://istio.io/latest/docs/reference/config/networking/sidecar/#WorkloadSelector

Definitely agree Service should behave this way though.

**Comment by miaoyq** [comment]
In my opinion, SE is different from others. The workloadSelector of SE is used to select WE, but not proxy. So the behavior should be similar to the service in k8s.

**Comment by hzxuzhonghu** [comment]
For sidecar authzpolicy Destinationrule, the empty or unset workload selector, we should keep using subsetof.

For SE selector, i think it is like k8s service, should be fixed

**Comment by howardjohn** [comment]
SE with empty selector then selects nothing (i.e. it's useless). We cannot
reject it now (breaking change) but can we at least add a warning in the
webhook?

Also for the others, they should use undefined selector instead of empty.
Again, can we warn in this case?

File: pkg/config/labels/instance.go [hunk 1]
```diff
@@ -73,6 +73,16 @@ func (i Instance) SubsetOf(that Instance) bool {
 	return true
 }
 
+// Selected is true if the label has same values for the keys.
+// if len(i) == 0, will return false. It is mainly used for service -> workload
+func (i Instance) Selected(that Instance) bool {
```
Comment (hzxuzhonghu) [inline pkg/config/labels/instance.go:78]: I would naming is `Match` here

**Comment by hzxuzhonghu** [comment]
Let me double check how k8s defining empty selector in service

**Comment by hzxuzhonghu** [comment]
// Route service traffic to pods with label keys and values matching this
	// selector. If empty or not present, the service is assumed to have an
	// external process managing its endpoints, which Kubernetes will not
	// modify. Only applies to types ClusterIP, NodePort, and LoadBalancer.
	// Ignored if type is ExternalName.
	// More info: https://kubernetes.io/docs/concepts/services-networking/service/
	// +optional
	// +mapType=atomic
	Selector map[string]string `json:"selector,omitempty" protobuf:"bytes,2,rep,name=selector"`

**Comment by hzxuzhonghu** [comment]
K8s implements selector in 

`staging/src/k8s.io/apimachinery/pkg/labels/selector.go`

We donot need to change istio CRD matching policy, only SE

File: pkg/config/labels/instance.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/config/labels/instance.go:78]: k8s uses `Macthes`

Comment (miaoyq) [reply pkg/config/labels/instance.go]: done

**Comment by miaoyq** [comment]
> We donot need to change istio CRD matching policy, only SE

Agree.

**Comment by howardjohn** [comment]
yes

**Review by hzxuzhonghu — APPROVED** [review]

File: releasenotes/notes/44159.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+releaseNotes:
+  - |
+    **Fixed** Fixes selector logic of service and endpoint.
```
Comment (hanxiaop) [inline releasenotes/notes/44159.yaml:6]: Can you provide more information about the updates to the selector logic mentioned in the release notes?

Comment (miaoyq) [reply releasenotes/notes/44159.yaml]: Done, PTAL :-)

File: releasenotes/notes/44159.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+releaseNotes:
+  - |
+    **Fixed** an issue of ServiceEntry and Service with undefined or empty workload selector. If the workload selector is undefined or empty,
```
Comment (hanxiaop) [inline releasenotes/notes/44159.yaml:6]:
```suggestion
    **Fixed** an issue where ServiceEntry and Service had undefined or empty workload selectors. If the workload selector is undefined or empty, 
```

**Review by howardjohn — COMMENTED** [review]
Please add validation as requested previously. Otherwise LGTM

**Review by howardjohn — CHANGES_REQUESTED** [review]
See above

**Comment by miaoyq** [comment]
> Please add validation as requested previously. Otherwise LGTM
@howardjohn Thx for remind,  I have added, PTAL

File: pkg/config/validation/validation.go [hunk 4]
```diff
@@ -1952,24 +1953,28 @@ func validateWorkloadSelector(selector *type_beta.WorkloadSelector) error {
 					fmt.Errorf("wildcard is not supported in selector: %q", fmt.Sprintf("%s=%s", k, v)))
 			}
 		}
+		if len(selector.MatchLabels) == 0 {
+			warning = fmt.Errorf("workload selector specified without labels," +
+				"should use undefined selector instead of empty") // nolint: stylecheck
```
Comment (hzxuzhonghu) [inline pkg/config/validation/validation.go:1958]: what is `undefined`, can be removed

File: pkg/config/validation/validation.go [hunk 5]
```diff
@@ -1939,8 +1939,9 @@ func ValidateControlPlaneAuthPolicy(policy meshconfig.AuthenticationPolicy) erro
 	return fmt.Errorf("unrecognized control plane auth policy %q", policy)
 }
 
-func validateWorkloadSelector(selector *type_beta.WorkloadSelector) error {
+func validateWorkloadSelector(selector *type_beta.WorkloadSelector) Validation {
 	var errs error
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1943]: Its preferred to have the style the rest of the package uses.

```
v := Validation{}
v = appendValidation(v, WrapWarning(some warning)))
v = appendValidation(v, fmt.Errorf(some error)))
return v.Unwrap()
```

File: pkg/config/validation/validation.go [hunk 6]
```diff
@@ -1939,8 +1939,9 @@ func ValidateControlPlaneAuthPolicy(policy meshconfig.AuthenticationPolicy) erro
 	return fmt.Errorf("unrecognized control plane auth policy %q", policy)
 }
 
-func validateWorkloadSelector(selector *type_beta.WorkloadSelector) error {
+func validateWorkloadSelector(selector *type_beta.WorkloadSelector) Validation {
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1942]: There is actually 2 types of WorkloadSelector, the other is validateAlphaWorkloadSelector which looks like already has this warning. So LGTM

**Review by howardjohn — APPROVED** [review]

File: pkg/config/validation/validation.go (hunk 5 above)
Comment (miaoyq) [reply pkg/config/validation/validation.go]: Updated.

**Pull request merged**